### PR TITLE
Add Java Solution to USACO-807

### PR DESCRIPTION
**Pull request by DarkHorse1997**
_If any of the below doesn't apply to this Pull Request, mark the checkbox as
done._

- [x] I have tested my code
- [x] I have followed the code style guidelines mentioned
      [here](https://usaco.guide/general/adding-solution/#code-conventions)
- [x] I have properly formatted the files according to the steps mentioned
      [here](https://usaco.guide/general/adding-solution#steps)

### CODE REVIEW

**Comment by nathangong** [comment]
I think the easier (and more intuitive solution) would just be casework, trying a --> b, a --> x --> y --> b, and a --> y --> x --> b.

**Comment by jessechoe10** [comment]
> I think the easier (and more intuitive solution) would just be casework, trying a --> b, a --> x --> y --> b, and a --> y --> x --> b.

If you look at official sol, it doesn't use any of the swapping. It just uses absolute value.

**Review by jessechoe10 — CHANGES_REQUESTED** [review]

**Comment by nathangong** [comment]
yeah by that i meant trying the three possible paths you can take

File: solutions/usaco-807.mdx [hunk 1]
```diff
@@ -43,5 +43,61 @@ print(ans, file = open('teleport.out', 'w'))
...
+
+        // Sort a,b in ascending order
+        int[] sortedlocations = sort(a, b);
+        a = sortedlocations[0];
+        b = sortedlocations[1];
+
+        // Sort x,y in ascending order
+        int[] sortedTeleportLocations = sort(x, y);
+        x = sortedTeleportLocations[0];
+        y = sortedTeleportLocations[1];
```
[the remark this answers is not in the document]
Comment (DarkHorse1997) [reply solutions/usaco-807.mdx]: Should I just do it like @nathangong mentioned? Find the minimum of the three possible distances.

File: solutions/usaco-807.mdx [hunk 2]
```diff
@@ -43,5 +43,61 @@ print(ans, file = open('teleport.out', 'w'))
...
+		if (a > b) {
+			return new int[] { a, b };
+		} else {
+			return new int[] { b, a };
+		}
+	}
+
+	public static void main(String[] args) throws IOException {
+		BufferedReader r = new BufferedReader(
+				new FileReader("teleport.in"));
```
Comment (SansPapyrus683) [inline solutions/usaco-807.mdx:64]: format like
```java
ReallyLongClassName instance = new ReallyLongClassName(
	[args]
);
```

File: solutions/usaco-807.mdx [hunk 3]
```diff
@@ -43,5 +43,61 @@ print(ans, file = open('teleport.out', 'w'))
...
+		// Distance needed to travel with teleportation
+		int teleportationDistance = Math.abs(x - a) + Math.abs(y - b);
+
+		// Result is the minimum of normal distance and teleportation distance
+		int result = Math.min(distance, teleportationDistance);
+
+		pw.println(result);
+
+		r.close();
+		pw.close();
```
Comment (SansPapyrus683) [inline solutions/usaco-807.mdx:96]:
```suggestion
		pw.close()
```

File: solutions/usaco-807.mdx [hunk 4]
```diff
@@ -43,5 +43,61 @@ print(ans, file = open('teleport.out', 'w'))
...
+
+public class Teleportation {
+	// Function to sort values in increasing order
+	static int[] sort(int a, int b) {
+		if (a > b) {
+			return new int[] { a, b };
+		} else {
+			return new int[] { b, a };
+		}
+	}
```
Comment (SansPapyrus683) [inline solutions/usaco-807.mdx:60]: this method is really questionable
i'd prefer you do something like
```java
if (a > b) {
	int temp = a;
	a = b;
	b = temp;
}
```
and something analogous for c & d

**Review by SansPapyrus683 — CHANGES_REQUESTED** [review]

File: solutions/usaco-807.mdx (hunk 3 above)
Comment (SansPapyrus683) [reply solutions/usaco-807.mdx:72]: not addressed

Comment (DarkHorse1997) [reply solutions/usaco-807.mdx:72]: I'm not sure what would be the benefit of removing `r.close()`. I am not using the reader anymore so why shouldn't I just call `close()` on it to free up resources.

Comment (SansPapyrus683) [reply solutions/usaco-807.mdx:72]: i mean at that point the program already ended

**Review by nathangong — APPROVED** [review]

**Review by jessechoe10 — APPROVED** [review]

**Pull request merged**